release-21.2: colexec: fixed nullableArgs projection operators

wenyihu6 opened the pull request:
Backport 2/2 commits from #84159.

/cc @cockroachdb/release

---

Previously, the `Concat` projection operators on arrays directly copy nulls into
the output without performing the actual projection whenever either argument was
null. This is incorrect because `Concat` operator can actually handle null arguments,
and we want the actual projection result instead. Since currently the only
projection operator that needs `nullableArgs == true` behaviour is
`ConcatDatumDatum`, this PR fixed this by adding a boolean field `nullableArgs`
to the logic for operators with `Datum`, `Datum`. If we later introduce another
projection operation that has nullableArgs == true, we should update this code
accordingly.

Fixes: https://github.com/cockroachdb/cockroach/issues/83091

Release note (bug fix): Fixed a bug in `Concat` projection operators on arrays
that gave output of nulls when the projection operator can actually handle null
arguments and may result in a non-null output.

----

Release justification: bug fix

wenyihu6 committed:
colexec: fixed nullableArgs projection operators

Previously, the `Concat` projection operators on arrays directly copy nulls into
the output without performing the actual projection whenever either argument was
null. This is incorrect because `Concat` operator can actually handle null arguments,
and we want the actual projection result instead. Since currently the only
projection operator that needs `nullableArgs == true` behaviour is
`ConcatDatumDatum`, this PR fixed this by adding a boolean field `nullableArgs`
to the logic for operators with `Datum`, `Datum`. If we later introduce another
projection operation that has nullableArgs == true, we should update this code
accordingly.

Fixes: https://github.com/cockroachdb/cockroach/issues/83091

Release note (bug fix): Fixed a bug in `Concat` projection operators on arrays
that gave output of nulls when the projection operator can actually handle null
arguments and may result in a non-null output.

wenyihu6 committed:
colexec: changed comments to template-time only

This commit changed some comments in `proj_const_ops_tmpl.go` and
`proj_non_const_ops_tmpl.go` to become template-time only.

Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/84967)
<!-- Reviewable:end -->

wenyihu6 committed:
sql: try fixing GitHub CI

This commit is trying to fix some GitHub CI failures. If it works, this should
be squashed into the previous commit.

yuzefovich commented:
I think it would be ok to not backport the fix to 21.2 since the bug seems relatively minor - the backport there is more difficult because some code has been moved around during 22.1 cycle. If you do decide to proceed with the backport, you need to use `make` for building and running tests locally.

wenyihu6 commented:
> I think it would be ok to not backport the fix to 21.2 since the bug seems relatively minor - the backport there is more difficult because some code has been moved around during 22.1 cycle. If you do decide to proceed with the backport, you need to use `make` for building and running tests locally.

Thanks! I will try again with `make`. If it's too hard, I will close this PR.

yuzefovich commented:
Oh, another command that you might need to run on 21.2 branch is `make execgen` - this is the way to regenerate all `.eg.go` files.
